### ISSUE TITLE
events that needs round trips, and the monsters revive

### ISSUE THREAD

**Issue by sujiniku** [clintbellanger/flare-game#707]
In some events, there are events that needs round trips with multiple map, but the enemies are revived because of map changing. So it is troublesome.

Especially in empyrean_campaign the enemies are also strong, and many enemys. So it is rather troublesome. 
For example, at the goblin ruby event, It is very troublesome to go back and forth between Kunma Lake.

So, if the hero defeat the enemy, please don't revive the enemys for a while even if the map changes.

**Comment by edipox** [comment]
I agree, I think some kind of timer for this would be very nice

**Comment by sujiniku** [comment]
In my idea, for each map, I record the last time I exited and the number of enemies I knocked down.

Again, when entering the map which already defeated the enemy from other map, it reduces the number of enemies that emerge, according to the number of enemies killed before.

Every time the enemies are resurrected. I assume that all enemies will be resurrected in 2 hours. It may seem long that it is two hours by roundtrip event, but it also includes the time such as checking equipment and buying and selling with traders, so I take a long time.

**Disadvantage.**
If the next map is long, there is a possibility that most enemies will be revived.

# grouping maps
So, furthermore, for each map, we make the map to group with related map.

In the case of round-trip events and associated maps, we ensure that enemies will not come back even if they move inside the round-trip map.

For this reason, it would be nice to group several maps within a group.

Just move between the maps in the group, it should not be revived even after some time.


To make this grouping easier, when you create a map, record where the map is located with map data. For example, about St.maria dungeon, record its dungeon to be in the "Lochport" area. 
If a map is inside the dungeon, in addition, we set in advance which building is in. That is, I set whether the map is a dungeon or a field (outside the building, outside the cave).

Different floors in one dungeon should be put together in one group.

It is troublesome that the enemy is resurrected when returning to the previous floor for reasons such as forgetting taking items.

Also, when entering  the cave on the field (or entering the fortress) and exiting from the cave, it is troublesome that the enemy is resurrected when going outside the cave.

So, such fields and caves or fortresses should be put together in one group.


When the gameover, let's not save the number of enemies that have been defeated.
Let's save the number of defeated enemies only when you move the map.

# How does the Hero earn XP?
About earning experience points, How does the player make it possible for the hero to earn experience points?

In past maps, it may be better to speed up the resurgence of enemy populations.

For example, if Hero reaches Kunma Lake, it is almost unnecessary to leave the enemies that have been killed by Lochport anymore.

However, even if you reach Kunma Lake, you may return to St.maria dungeon due to forgetting to pick up items, so bother to resurrect the enemies of st.maria dungeon will make the player uncomfortable.

On a map with two warp zones before the last, it may be better to restore in about an hour.

Or, with reference to the level of the hero, it may be better to hasten the enemy 's resurgence speed by itself, on a map with weak enemies.

At the last dungeon, if this is the case, since enemies are not likely to recover easily , there is a possibility that it will be difficult to earn experience point. 
So I think that it is good to speed up enemy's revive speed only by the last dungeon so much.

**Comment by Reylak** [comment]
Adding to the mob respawn problem on map transition, there is the dropped loot issue.

When running around kiting enemies close to the map entrance, it's not unusual for me to click on the entrance and leave the map. This makes every mob respawn, voiding my effort to clean the place, and it also destroys all unpicked loot.

So really, for me areas should **not** reset upon leaving. If the challenge is indeed to make roundtrips, an event can spawn more enemies on the way back. If the player is farming a zone, the game can reset the map on entering Hyperspace, or on player's command like holding Maj while entering the map (I believe PoE does that).

**Comment by sujiniku** [comment]
I think that, Flare needs to save the state of the map for all maps.

This will greatly increase the size of the save data, but it would maybe not impossible. For example, in commercially available history simulation games or war simulation games etc,these game saves dozens of nation stats.

Well,I thinkthat Flare does not need to preserve the monster 's HP.
It is necessary to save for each monster whether each monster on that map is alive or dead.

When hero moves the map in the state that the monster is alive, it is easy and natural to recover the alive monster's HP in its entirety.

For dropped items, Flare also needs to save where it is dropping and what is dropped.
I think that, Instead of creating saved data after entering the map, it is better that Flare should create statistics data for that map since the beginning (since New game).
If the Hero enters the map, which map stats to overwrite is changed to new map .

**Comment by Reylak** [comment]
I disagree with saving any map state upon leaving the game (except unlocked path such as at the end of dungeons, or any resolved environmental puzzle).

Any hack'n'slash I know of, resets its maps (procedurally generated or not) when leaving the game. You are not supposed to totally and permanently wipe out all the maps. Plus, you most often need to farm in order to pass the next map. And you _want_ to farm in order to complete this marvelous item set that gives you +100MP et +25% mental ;)

So to summarize, my suggestion in this matter is:

 * keep any map's state in memory (possibly with a time limit, as suggested by @sujiniku [previously](https://github.com/clintbellanger/flare-game/issues/707#issuecomment-372536969) because that may lead to a huge memory usage), this includes dropped items and killed monsters (but monsters' HP are reset)
 * allow for the player to manually reset a map upon entering it, by holding the Maj button or any other solution

This solves @sujiniku's problem of boring round-trips, and my issue of wild map resets when misclicking on map exit.

I don't know what part of this is actually engine-related rather than related to the game itself though...
